crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -498,21 +498,57 @@ impl<'db> TypeInferenceBuilder<'db> {
                             format_args!(
                                 "Invalid class base with type `{}` (all bases must be a class, `Any`, `Unknown` or `Todo`)",
                                 base_ty.display(self.db)
-                            )
+                            ),
                         );
                     }
-                },
-                MroErrorKind::UnresolvableMro{bases_list} => self.diagnostics.add(
+                }
+                MroErrorKind::UnresolvableMro { bases_list } => self.diagnostics.add(
                     class.node(self.db).into(),
                     "inconsistent-mro",
                     format_args!(
                         "Cannot create a consistent method resolution order (MRO) for class `{}` with bases list `[{}]`",
                         class.name(self.db),
                         bases_list.iter().map(|base| base.display(self.db)).join(", ")
-                    )
+                    ),
                 )
             }
         }
+
+        let class_definitions = self
+            .types
+            .declarations
+            .values()
+            .filter_map(|ty| ty.into_class_literal())
+            .map(|class_ty| class_ty.class);
+
+        let invalid_metaclasses = class_definitions.filter_map(|class| {
+            class
+                .try_metaclass(self.db)
+                .err()
+                .map(|metaclass_error| (class, metaclass_error))
+        });
+
+        for (class, metaclass_error) in invalid_metaclasses {
+            match metaclass_error.reason() {
+                MetaclassErrorKind::Conflict {
+                    metaclass1,
+                    metaclass2
+                } => self.diagnostics.add(
+                    class.node(self.db).into(),
+                    "conflicting-metaclass",
+                    format_args!(
+                        "The metaclass of a derived class (`{}`) must be a subclass of the metaclasses of all its bases, but `{}` and `{}` are not compatible",
+                        class.name(self.db),
+                        Type::ClassLiteral(*metaclass1).display(self.db),
+                        Type::ClassLiteral(*metaclass2).display(self.db),
+                    ),
+                ),
+                MetaclassErrorKind::CyclicDefinition => {
+                    // TODO(charlie): When diagnostics are deduplicated, add a `cyclic-class-def`
+                    // diagnostic, equivalent to the above.
```

Comment:
I don't think we need this TODO? I don't think we ever have plans to de-duplicate diagnostics in a way that would de-deduplicate two diagnostics that we explicitly raise separately. And I think a cyclic class definition is a cyclic class definition; I don't think there are cases where the MRO cycle detection would fail to detect a cycle but we'd detect it here.
```suggestion
                    // Cyclic class definition diagnostic will already have been emitted above in MRO calculation.
```

An interesting question here: if this is true, do we really need cycle detection implemented separately in both MRO and metaclass calculation? It seems inefficient, both in code to maintain and at runtime. Could we instead hoist the "is cyclically defined" check (which is already done as a totally separate function in MRO calculation) out into this function, and if a class is cyclically defined, we emit a diagnostic and just `specify` (this is a Salsa feature) the result of its MRO and metaclass queries as Unknown, and never try to calculate them. Then the MRO and metaclass calculation code could both take non-cyclic as an assumed invariant.

This might be a big enough change to do as a separate PR, but it seems worth doing to me, unless I'm missing something. Curious what @AlexWaygood and @MichaReiser think.

Reply:
It's definitely an interesting question where cycle detection should go. In an earlier version of my MRO PR, I had cycle-detection as a separate Salsa query, which was called by the MRO-resolution logic but was not itself part of the MRO resolution logic (my PR did this until https://github.com/astral-sh/ruff/pull/14027/commits/e1f1874e9f195c2953f155762668f362f14145aa, which moved changed it to the current implementation of cycle detection in MROs).

I'd vote for leaving it as @charliermarsh has it for now, though, and investigating whether we can consolidate/optimise the cycle-detection logic as a followup.

Reply:
https://github.com/astral-sh/ruff/issues/14141